src/framework/components/element/markup.js

```diff
@@ -0,0 +1,417 @@
+
+        // returns the buffer for the last returned token
+        buf: function () {
+            return this._buf;
+        },
+
+        // returns the index of end of the last successful token extraction
+        last: function () {
+            return this._last;
+        },
+
+        // return the error message
+        error: function () {
+            return this._error;
+        },
+
+        // print the scanner output
+        debugPrint: function () {
+            var tokenStrings = ["EOF", "ERROR", "TEXT", "OPEN_BRACKET", "CLOSE_BRACKET", "EQUALS", "STRING", "IDENTIFIER", "WHITESPACE"];
+            var token = this.read();
+            var result = "";
+            while (true) {
+                result += (result.length > 0 ? "\n" : "") +
+                            tokenStrings[token] +
+                            " '" + this.buf().join("") + "'";
+                if (token === this.EOF_TOKEN || token === this.ERROR_TOKEN) {
+                    break;
+                }
+                token = this.read();
+            }
+            return result;
+        },
+
+        // read the next token from the input stream and return the token
+        _read: function () {
+            this._buf = [];
+            if (this._eof()) {
+                return this.EOF_TOKEN;
+            }
+            return (this._mode === "text") ? this._text() : this._tag();
+        },
+
+        // read text block until eof or start of tag
+        _text: function () {
+            while (true) {
+                switch (this._cur) {
+                    case null:
+                        // reached end of input
+                        return (this._buf.length > 0) ? this.TEXT_TOKEN : this.EOF_TOKEN;
+                    case "[":
+                        // start of tag mode
+                        this._mode = "tag";
+                        return (this._buf.length > 0) ? this.TEXT_TOKEN : this._tag();
+                    case "\\":
+                        // handle escape sequence
+                        this._next();           // skip \
+                        switch (this._cur) {
+                            case "[":
+                                this._store();
+                                break;
+                            default:
+                                // if we don't recognize the escape sequence, output
+                                // the slash without interpretation and continue
+                                this._output("\\");
+                                break;
+                        }
+                        break;
+                    default:
+                        this._store();
+                        break;
+                }
+            }
+        },
+
+        // read tag block
+        _tag: function () {
+            while (true) {
+                switch (this._cur) {
+                    case null:
+                        this._error = "unexpected end of input reading tag";
```

Comment:
Question about all the errors in this file - I'm wondering if users will understand what's going on if they see one of those errors in their console. Maybe we should be more specific that this is about color tags in text elements?

Reply:
I guess you have a more descriptive error below so that should be fine...

Reply:
Yes, these describe the specific errors. They'll be packaged up into a nice little message which states markup.